plugins/image-prioritizer/helper.php

```diff
@@ -77,6 +77,142 @@ function image_prioritizer_register_tag_visitors( OD_Tag_Visitor_Registry $regis
+}
+
+/**
+ * Filters the list of Optimization Detective extension module URLs to include the extension for Image Prioritizer.
+ *
+ * @since n.e.x.t
+ *
+ * @param string[]|mixed $extension_module_urls Extension module URLs.
+ * @return string[] Extension module URLs.
+ */
+function image_prioritizer_filter_extension_module_urls( $extension_module_urls ): array {
+	if ( ! is_array( $extension_module_urls ) ) {
+		$extension_module_urls = array();
+	}
+	$extension_module_urls[] = add_query_arg( 'ver', IMAGE_PRIORITIZER_VERSION, plugin_dir_url( __FILE__ ) . image_prioritizer_get_asset_path( 'detect.js' ) );
+	return $extension_module_urls;
+}
+
+/**
+ * Filters additional properties for the element item schema for Optimization Detective.
+ *
+ * @since n.e.x.t
+ *
+ * @param array<string, array{type: string}> $additional_properties Additional properties.
+ * @return array<string, array{type: string}> Additional properties.
+ */
+function image_prioritizer_add_element_item_schema_properties( array $additional_properties ): array {
+	$additional_properties['lcpElementExternalBackgroundImage'] = array(
+		'type'       => 'object',
+		'properties' => array(
+			'url'   => array(
+				'type'      => 'string',
+				'format'    => 'uri',
+				'required'  => true,
+				'maxLength' => 500, // Image URLs can be quite long.
+			),
+			'tag'   => array(
+				'type'      => 'string',
+				'required'  => true,
+				'minLength' => 1,
+				'maxLength' => 25, // The longest HTML tag name is 10 characters (BLOCKQUOTE and FIGCAPTION). This maxLength accounts for possible Custom Elements that are even longer.
+				'pattern'   => '^[a-zA-Z0-9\-]+$', // Technically emoji can be allowed in a custom element's tag name, but this is not supported here.
+			),
+			'id'    => array(
+				'type'      => array( 'string', 'null' ),
+				'maxLength' => 100, // A reasonable upper-bound length for a long ID.
+				'required'  => true,
+			),
+			'class' => array(
+				'type'      => array( 'string', 'null' ),
+				'maxLength' => 500, // There can be a ton of class names on an element.
+				'required'  => true,
+			),
+		),
+	);
+	return $additional_properties;
+}
+
+/**
+ * Validates that the provided background image URL is valid.
+ *
+ * @since n.e.x.t
+ *
+ * @param bool|WP_Error|mixed  $validity   Validity. Valid if true or a WP_Error without any errors, or invalid otherwise.
+ * @param OD_Strict_URL_Metric $url_metric URL Metric, already validated against the JSON Schema.
+ * @return bool|WP_Error Validity. Valid if true or a WP_Error without any errors, or invalid otherwise.
+ */
+function image_prioritizer_filter_store_url_metric_validity( $validity, OD_Strict_URL_Metric $url_metric ) {
+	if ( ! is_bool( $validity ) && ! ( $validity instanceof WP_Error ) ) {
+		$validity = (bool) $validity;
+	}
+
+	$data = $url_metric->get( 'lcpElementExternalBackgroundImage' );
+	if ( ! is_array( $data ) ) {
+		return $validity;
+	}
+
+	$r = wp_safe_remote_head(
+		$data['url'],
```

Comment:
The `wp_safe_remote_head()` function passes the supplied URL through `wp_http_validate_url()`, but that doesn't necessarily mean the request is safe. I'm thinking there should perhaps be an allowlist of hosts that this is allowed to connect to. Maybe we query for the first image attachment and pass it into `wp_get_attachment_image_url()` to then extract the hostname from the image. This would be better than having to try to come up with a list of all image CDNs that are used by plugins.

The list of allowed hosts could also be sent to the frontend so as to avoid sending the invalid URL in the first place. See related todo:

https://github.com/WordPress/performance/blob/bc81588943e835707d70948bdd6a3c817c948a25/plugins/optimization-detective/detect.js#L361-L366

Then again, would be a shape to continuously reject all of the URL Metrics in the case that an unexpected host is being used for a given background image. In that case, perhaps we should allow `OD_URL_Metric` to be mutable, so in the case where an invalid host is provided this callback could instead do:

```php
$url_metric->set( 'lcpElementExternalBackgroundImage', null );
```

or rather

```unset
$url_metric->unset( 'lcpElementExternalBackgroundImage' );
```

When mutating an instance of `OD_URL_Metric` it should re-trigger the validation so that the mutation would cause the JSON Schema to be made invalid that it would throw an exception.

Also, whenever mutation happens then it should call a new `$this->group->clear_cache()` method which empties out `$group->result_cache` and which would in turn call `$collection->clear_cache()`.

Reply:
So this `od_store_url_metric_validity` filter could be used in two ways:

1. A filter callback could return `false` or a non-empty `WP_Error` instance to block the URL Metric from being stored at all.
2. A filter callback could mutate the supplied `OD_URL_Metric` object to apply custom sanitizations or validation constraints which then would end up getting persisted in storage.

Reply:
I think this would be good to open a separate issue for. This PR is already quite big as is, and this could be an iterative enhancement.

Reply:
I went ahead and reverted the new filter from this PR and moved it into a follow-up one: https://github.com/WordPress/performance/pull/1713